src/Installer.php

```diff
@@ -160,8 +170,31 @@ private function installPuli()
 
         $httpClient = new HttpClient($this->disableTls, $this->cafile);
 
+        $versionUrl = sprintf(
+            static::VERSION_API_URL,
+            $this->disableTls ? 'http' : 'https'
+        );
+
+        for ($retries = 3; $retries > 0; --$retries) {
+            if (!($versions = json_decode($httpClient->download($versionUrl), true))) {
+                continue;
+            }
+
+            break;
+        }
+
+        if (0 === $retries) {
+            $this->error('The download failed repeatedly, aborting.');
+
+            return 1;
+        }
+
+        if (null === ($this->version = $this->matchVersion($versions))) {
```

Comment:
`getLatestUnstableVersion` could also be inline
But in my opinion the helper methods improve code readability.